action_server/src/action_server/actions/example_action.py

```diff
@@ -0,0 +1,133 @@
+        required in the semantics of the task, and what information from previous tasks is required.
+        """
+
+        """
+        You can set the required robot skills here. During the configure step, the given robot object is checked for
+        their availability.
+        """
+        self._required_skills = ['head', 'hmi', 'base']
+
+        """
+        Similarly, you can set the required keys in the knowledge. Their availability is also checked in the configure
+        step.
+        """
+        self._required_passed_knowledge = []
+
+        """
+        Similarly, you can set the required keys in the semantics. Their availability is also checked in the configure
+        step. However, a missing field means that the client may want to add this information by posing a follow-up
+        question. Therefore, this is a dict mapping from the required fields to the prompts requesting the user for
+        the specific missing information.
+
+        If information is missing, the json path to the missing field is returned to the client, so the client knows
+        which field to add.
+        """
+        self._required_field_prompts = {'object': " What object would you like me to manipulate? "}
+
+        """
+        Objects that should be passed on to a next action should be instantiated here.
+        """
+        self._object = Entity()
+
+    def _configure(self, robot, config):
+        """
+        :param robot: The robot to use for this action
+        :param config: The configuration of the action, containing fields semantics and knowledge
+        :return: ConfigurationResult
+
+        The configure step is typically done just after the command is given. It is intended to check if the task makes
+        any sense. For example, a task 'Take it to the bridge.', does not make sense as 'it' should refer to something.
+        In another example, 'Find Ava', we may want to search differently than when the task is to find an object. All
+        such checking can be done in the configure step.
+
+        If a command is composed out of multiple actions (go to the kitchen and find the coffee), all actions in the
+        command are first configured before the first is started. This is done to make sure the complete command makes
+        sense, and the robot will not find out it doesn't make sense halfway through the execution.
+
+        This does mean that any resulting information from earlier actions in the command is not available in the world
+        model yet, in this stage.
+        """
+
+        """
+        The common knowledge is available through self._knowledge
+        """
+        if config.semantics['object']['type'] in self._knowledge.object_categories:
+            category = config.semantics['object']['type']
+        else:
+            self._config_result.message = " The specified object type is not an object I know. "
+            return
+
```

Comment:
What bothers me here is that this is ``ExampleAction`` right, ie it knows what to expect in ``config.semantics``. So why does the code use these ``config.semantics['object']['type']`` phrases all over the place? Instead, decode the semantics into a few data fields of the object, which then gives code like ``category = self.sem_object_type`` (Not a very good name for the variable, but I hope you get the idea.)

Similarly,perhaps add a ``def set_result_message(msg, success)`` method?

Reply:
It would be nice to do this using some interface specification of the action. What I mean is that `ExampleAction` could specify its interface using a json schema. A parser could use this to construct the object from the configuration dict before any of the fields are accessed. This parser can also throw errors of any missing required information. The implementation's `_configure` method could use this output and link it to some semantic meaning i.e. the `configuration_result. message`.

All this is future work for now. I first want to document the current state, then add this sort features later.